core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -2200,4 +2197,222 @@ public void testDocStats() throws IOException {
         assertEquals(0, docStats.getDeleted());
         assertEquals(numDocs, docStats.getCount());
     }
+
+    public void testDoubleDelivery() throws IOException {
```

Comment:
I'm a bit confused by this test compare to the other two - what do we test here that the other don't and in what scenarios that can happen? I feel like I'm missing something